lib/rules/redux-no-bound-selectors.js

```diff
@@ -15,59 +15,152 @@ const BIND_ERROR_MESSAGE = "Don't bind functions within `connect`." + REF;
+				'within a function passed as the first argument to `connect`.',
 			recommended: true
 		},
 		schema: []
 	},
 	create: function( context ) {
-		let hasStateInClosure = false;
+		/*
+		 * An array of tuples of shape
+		 *
+		 *   ( innerNode, outerNode, reportMessage )
+		 *
+		 * where
+		 *
+		 *   - `innerNode` refers to the offending node, i.e. the Identifier or
+		 *   [Arrow]FunctionExpression where binding/instantiation occurs
+		 *
+		 *   - `outerNode` refers to the closest [Arrow]FunctionExpression
+		 *   ancestor that introduces `state` into the closure
+		 *
+		 *   - `reportMessage` is the error string to be passed to eslint's
+		 *   `context.report`
+		 *
+		 * Since we need to find all occurrences of `connect` before we can
+		 * tell which state-aware functions are actually forbidden from
+		 * including binding/instantiation of new functions, we progressively
+		 * store them in this collection.
+		 *
+		 */
+		const reports = [];
+
+		/*
+		 * A collection of bodies of [Arrow]FunctionExpressions that ever
+		 * become the first argument of `connect`. In a typical module, this
+		 * array will contain 0 or 1 elements.
+		 *
+		 * In Program:exit, any element in `reports` whose `outer` property is
+		 * contained in `nodesToReportOn` will yield an actual eslint report.
+		 */
+		const nodesToReportOn = [];
+
+		/*
+		 * Naming is hard. ¯\_(ツ)_/¯
+		 *
+		 * Refers to a body of a function that introduces `state` (via its
+		 * arguments) into the subsequent closures.
+		 *
+		 * When traversing the AST, if this value is falsy we can assume that
+		 * we are not in a closure whose broader scope contains `state`.
+		 */
+		let stateProviderFunctionBody;
 
 		function onFunctionEnter( node ) {
-			if ( hasStateInClosure ) {
-				context.report( node, FUNC_ERROR_MESSAGE );
+			if ( stateProviderFunctionBody ) {
+				reports.push( {
+					inner: node,
+					outer: stateProviderFunctionBody,
+					message: FUNC_ERROR_MESSAGE,
+				} );
 				return;
 			}
 
-			if ( isStateInArgs( node ) && ! isCallbackOfSetState( node ) ) {
-				hasStateInClosure = true;
+			if ( isStateInArgs( node ) ) {
+				stateProviderFunctionBody = node.body;
 			}
 		}
 
 		function onFunctionExit( node ) {
-			if ( hasStateInClosure && isStateInArgs( node ) ) {
-				hasStateInClosure = false;
+			if ( stateProviderFunctionBody && isStateInArgs( node ) ) {
+				stateProviderFunctionBody = null;
+			}
+		}
+
+		function onIdentifierEnter( node ) {
+			if ( stateProviderFunctionBody && node.name === 'bind' ) {
+				reports.push( {
+					inner: node,
+					outer: stateProviderFunctionBody,
+					message: BIND_ERROR_MESSAGE,
+				} );
 			}
 		}
 
 		return {
+			'Program:exit': function() {
+				reports
+					.filter( ( { outer } ) =>
+						contains( nodesToReportOn, outer ) )
+					.forEach( ( { inner, message } ) =>
+						context.report( inner, message ) );
+			},
+
+			CallExpression: function( node ) {
+				if ( node.callee.name !== 'connect' ) {
```

Comment:
Doing a quick search in Calypso, I think we can assume that it's probably from react-redux.